openhtf/core/__init__.py

```diff
@@ -107,18 +108,24 @@ def _thread_proc(self):
     """Handles one whole test from start to finish."""
     with contextlib.ExitStack() as exit_stack:
       # Top level steps required to run a single iteration of the Test.
+      self.test_state = test_state.TestState(self._test_descriptor)
+      executor = phase_executor.PhaseExecutor(self.test_state)
 
       # Any access to self._exit_stack must be done while holding this lock.
       with self._lock:
-        # Initial setup of exit stack and final cleanup of attributes.
         self._exit_stack = exit_stack
-
-      self.test_state = test_state.TestState(self._test_descriptor)
-      # Wait here until the test start trigger returns a DUT ID.  Don't hold
-      # self._lock while we do this, or else calls to stop() will deadlock.
-      # Create plugs while we're here because that may also take a while and
-      # we don't want to hold self._lock while we wait.
-      self.test_state.mark_test_started(self._wait_for_test_start())
+        exit_stack.callback(executor.stop)
```

Comment:
hm, you probably still want to `stop()` the executor *before* tearing down plugs.  technically there may still be a phase running if the executor isn't stopped.  This means pushing it onto the `exit_stack` after `tear_down_plugs` below.